### PR TITLE
Fix Federation: Google Cloud DNS provider does not handle missing dom…

### LINKED ISSUE #51139: Federation: Google Cloud DNS provider does not handle missing domains correctly
/kind bug

**What happened**:
walteraa [6:55 AM] 
Hi, The controller manager of the federation using Google cloud DNS is crashing, the log showed is that:

```I0822 13:17:33.997888       1 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:jwt.jwtSource{ctx:(*context.emptyCtx)(0xc42001c550), conf:(*jwt.Config)(0xc420288e00)}, mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(nil)}
I0822 13:17:33.998181       1 clouddns.go:100] Successfully got DNS service: &{0xc420e19b00 https://www.googleapis.com/dns/v1/projects/  0xc42000e840 0xc42000e848 0xc42000e850 0xc42000e858}
E0822 13:17:35.621941       1 dns.go:97] Failed to retrieve DNS zone: error querying for DNS zones: googleapi: got HTTP response code 404 with body: Not Found
F0822 13:17:35.622009       1 controllermanager.go:144] Failed to start service dns controller: error querying for DNS zones: googleapi: got HTTP response code 404 with body: Not Found
```

the hyperkube command is that:
```/hyperkube
      federation-controller-manager
      --dns-provider=google-clouddns
      --federation-name=fed-keepalived
      --kubeconfig=/etc/federation/controller-manager/kubeconfig
      --master=https://fed-keepalived-apiserver
      --zone-name=kfed.walteralves.me.
```

I have created this `kfed.walteralves.me.` in the Google DNS

Someone knows why the DNS zone query is getting the 404 error?? :disappointed:

**What you expected to happen**:

It should have created the missing domain.

**How to reproduce it (as minimally and precisely as possible)**:

As per above.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.4
- Cloud provider or hardware configuration**: OpenStack (irrelevant)
- OS (e.g. from /etc/os-release): irrelevant
- Kernel (e.g. `uname -a`): irrelevant
- Install tools: irrelevant
- Others: irrelevant

### PR DESCRIPTION
**Pull request by shiywang**
Fixes https://github.com/kubernetes/kubernetes/issues/51139
Not sure how to fix provider code, submit a pr to google could dns or something ?
also I'm still learning how to setup a federation cluster, will test my code after it finished : )
/assign @quinton-hoole 
@kubernetes/sig-federation-bugs @kubernetes/huawei

### CODE REVIEW

**Comment by quinton-hoole** [comment]
Thanks @shiywang !  

The way you would fix the provider code is by making sure that the following return an empty list rather than an error when not found:

https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/providers/aws/route53/zones.go#L33

https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/providers/aws/route53/zones.go#L33

https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/providers/coredns/zones.go#L32

And update the documentation for provider.Zones.List() accordingly.

By the way, I'm not sure that your current code will work correctly on all providers, as it's not guaranteed that they will all return the error type and code you are assuming in this PR's code.

It might be safer to fix each of the providers, to make sure they all return the right result (i.e. empty set and no error if the result set is empty).  The current API definition (provider.Zones) is not specific enough to be able to assume that you'll receive a errors.APIStatus type back from dnsprovider.Zones.List().

**Comment by shiywang** [comment]
@quinton-hoole thanks for this detail elaboration, I already chage those provider.Zones.List() to return an emptyList and nil if operation failed, since you said there's not guaranteed return error code is 404, so I assume as long as list operation failed, we return an empty list and no error, is that right ? ptal

**Comment by quinton-hoole** [comment]
@shiywang Yes, that's right.  I'll take a look ASAP, although probably not for another week at least, as I'm travelling.

File: federation/pkg/dnsprovider/dns.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ type Interface interface {
 }
 
 type Zones interface {
-	// List returns the managed Zones, or an error if the list operation failed.
+	// List returns the managed Zones, or an empty list if the list operation failed.
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/dns.go]: Please change to:

// List returns the (possibly empty) list of managed Zones, or an error if the list operation failed.

File: federation/pkg/dnsprovider/providers/aws/route53/zones.go [hunk 2]
```diff
@@ -34,15 +34,12 @@ func (zones Zones) List() ([]dnsprovider.Zone, error) {
 	var zoneList []dnsprovider.Zone
 
 	input := route53.ListHostedZonesInput{}
-	err := zones.interface_.service.ListHostedZonesPages(&input, func(page *route53.ListHostedZonesOutput, lastPage bool) bool {
+	zones.interface_.service.ListHostedZonesPages(&input, func(page *route53.ListHostedZonesOutput, lastPage bool) bool {
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/aws/route53/zones.go]: You can't just ignore the error here.  If the route53 API returns an error to indicate that no zones exist, you need to return an empty list and no error.  But for other errors (for example authentication errors, or unreachable API) you need to return that error.

File: federation/pkg/dnsprovider/providers/google/clouddns/zones.go [hunk 3]
```diff
@@ -30,12 +30,13 @@ type Zones struct {
 }
 
 func (zones Zones) List() ([]dnsprovider.Zone, error) {
+	var zoneList []dnsprovider.Zone
 	response, err := zones.impl.List(zones.project()).Do()
 	if err != nil {
-		return nil, err
+		return zoneList, nil
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/google/clouddns/zones.go]: Same comment here as above.  Only suppress the error if the call succeeded but returned an empty list.  For all other types of errors, you need to return them, not discard them here.

**Review by ghost — COMMENTED** [review]
This PR is still not correct.  Please see comments.  Also have you checked the other DNSProvider implementations (e.g. coreos)?

**Comment by irfanurrehman** [comment]
**FYI**
sig-multicluster is planning a cutover of code from core kubernetes to its own repo (k8s.io/kubernetes/federation), and the deadline to merge PRs was posted [here](https://groups.google.com/forum/#!topic/kubernetes-sig-multicluster/DW6odiufiBs). 
@shiywang this seems to be waiting for the fix of the review comments; Will you be able to finish this by this weekend? (If not so, you will need to raise the fix against the new repo once the cutover is complete; sometime next week).
cc @kubernetes/sig-multicluster-misc

**Comment by shiywang** [comment]
@quinton-hoole @irfanurrehman sorry for my very late response, I updated as suggestion, for coredns, i think it doesn't return any error, so no need to change I think https://github.com/kubernetes/kubernetes/blob/2807e8050e27a7f3c958a63592340018dc359618/federation/pkg/dnsprovider/providers/coredns/zones.go#L37.

also for cloud dns return error message, I haven't had a chance to setup an clouddns to test properly since I'm an individual contributor. @irfanurrehman do you have any ideas how do i test the error messages ?

**Pull request merged**